### PR TITLE
Utilize Eigen math library

### PR DESCRIPTION
**Pull request by beholdnec**
This PR imports the Eigen math library (version 3.3.4) and uses it to perform matrix math in some parts of VideoCommon and WiimoteEmu, replacing some of our older ad-hoc matrix code. Our code becomes more concise, and we gain a stable, high-quality math library that we can use in the future.

Eigen was chosen because:
- It is licensed under the MPL2 which is compatible with our license.
- It consists purely of headers and does not rely on third-party libraries.
- It has powerful matrix classes -- matrices are generic over type and size -- and in fact, a vector in Eigen is just a matrix with one column.
- It is reputed to be both fast and easy to use.

### CODE REVIEW

File: Source/Core/Common/MathUtil.h [hunk 1]
```diff
@@ -187,6 +189,21 @@ struct Rectangle
   }
 };
 
+// Vectors are matrices with one column.
+typedef Eigen::Matrix<float, 2, 1, Eigen::DontAlign> Vector2f;
```
Comment (lioncash) [inline Source/Core/Common/MathUtil.h]:
```cpp
using Vector2f = ...
```

etc.

File: Source/Core/InputCommon/ControllerEmu/ControllerEmu.h [hunk 2]
```diff
@@ -14,7 +14,11 @@
 
 class ControllerInterface;
 
-#define sign(x) ((x) ? (x) < 0 ? -1 : 1 : 0)
+template <typename T>
```
Comment (lioncash) [inline Source/Core/InputCommon/ControllerEmu/ControllerEmu.h]: While a good change, this doesn't seem to be related to using Eigen as far as I can tell and should probably go in its own commit.

**Comment by Helios747** [comment]
Is cleaner math worth another dep?

File: Source/Core/InputCommon/ControllerEmu/ControllerEmu.h (hunk 2 above)
Comment (beholdnec) [reply Source/Core/InputCommon/ControllerEmu/ControllerEmu.h]: This change is needed to fix compile errors where "sign" is used as an identifier.

**Comment by JosJuice** [comment]
The dependency is too large for the benefit we get, in my opinion.

**Pull request closed without merging**